---
title: "A Brief Tutorial of cisTopic"
author: "YOUR_NAME"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
---


# What is cisTopic? 

cisTopic is an R/Bioconductor package for the simulataneous identification of *cis-regulatory topics* and cell states from single cell epigenomics data. cisTopic relies on an algorithm called Latent Dirichlet Allocation (LDA), a robust Bayesian method used in text mining to group documents addressing similar topics and related words into topics. Interestingly, this model has a series of assumptions that are fulfilled in single-cell epigenomics data, such as non-ordered features (*'bag of words'*) and the allowance of overlapping topics (i.e. a regulatory region can be co-accessible with different other regions depending on the context, namely, the cell type or state). 

cisTopic uses LDA with a collapsed Gibbs sampler (Griffiths & Steyvers, 2004), where each region in each cell is assigned to a topic based on (1) to which topic the region is assigned in other cells and (2) to which topics the regions are assigned in that cell. After a number of iterations through the data set, these assignments are used to estimate the probability of a region belonging to a cis-regulatory topic (region-topic distribution) and the contributions of a topic within each cell (topic-cell distribution). These distributions can in turn be used to cluster cells and identify cell types, and to analyse the regulatory sequences in each topic. 

cisTopic consists of 4 main steps: (1) generation of a binary accessibility matrix as input for LDA; (2) LDA and model selection; (3) cell state identification using the topic-cell distributions from LDA and (4) exploration of the region-topic distributions. 

![Figure 1. cisTopic workflow. The input for cisTopic is a binary accessibility matrix. This matrix can be formed from single-cell BAM files and a set of genome-wide regulatory regions (e.g., from peak calling on the bulk or aggregate data). Next, Latent Dirichlet Allocation (LDA) is applied on the binary accessibility matrix to obtain the topic-cell distributions (contributions of each topic per cell) and the region-topic distributions (contributions of each region to a topic). Note that a region can contribute to more than one topic (represented by the purple peaks). Finally, the topic-cell distributions are used for dimensionality reduction (e.g. PCA, tSNE, diffusion maps) and clustering to identify cell states, while the region-topic distributions can be used to predict the regulatory code underlying the topic.](./10x_example/Fig1.png) 


```{r, eval=T, results='asis', message=F}

suppressMessages(library(cisTopic))
suppressMessages(library(plyr))
suppressMessages(library(tidyverse))
```


# Running cisTopic
## Input data

Some steps in this tutorial might take a few minutes to run, as reference we mention the running time for this dataset and settings in our system. Your actual running time will depend on your computer and dataset. In this tutorial, we will run cisTopic on 5,335 Peripheral blood mononuclear cells (PBMCs) from a healthy donor, with 97k potential regulatory regions.

The cisTopic object can be initialized for 10X data can either from **(a)** the count matrix produced by CellRanger ATAC; or **(b)** CellRanger ATAC fragments file and a bed file with the candidate regulatory regions. If the user prefers to define peaks in a different way from CellRanger ATAC, we recommend to use option **b**. 

The cisTopic object contains all the information and outputs from the analysis. For more information, run:


```{r}
cisTopicObject <- readRDS('cisTopicObject_pbmc.Rds')
```

```{r, eval=FALSE, results='asis'}
?`cisTopic-class`
```

For initializing the cisTopic object:

* a.  Starting from the **CellRanger ATAC count matrix** [Reference running time (5k cells, 97k regions): 1,5 min]
```{r, eval=F}
pathTo10X <- './10x_example/'
data_folder <- paste0(pathTo10X, 'filtered_peak_bc_matrix')
metrics <- paste0(pathTo10X, 'atac_v1_pbmc_5k_singlecell.csv')
cisTopicObject <- createcisTopicObjectFrom10Xmatrix(data_folder, metrics,  project.name='5kPBMCs')
```

## Building the models

The next step in the cisTopic workflow is to use Latent Dirichlet Allocation (LDA) for the modelling of cis-regulatory topics. LDA allows to derive, from the original high-dimensional and sparse data, (1) the probability distributions over the topics for each cell in the data set and (2) the probability distributions over the regions for each topic (Blei et al., 2003). These distributions indicate, respectively, how important a regulatory topic is for a cell, and how important regions are for the regulatory topic. Here, we use a collapsed Gibbs sampler (Griffiths and Steyvers, 2004), in which we assign regions to a certain topic by randomly sampling from a distribution where the probability of a region being assigned to a topic is proportional to the contributions of that region to the topic and the contributions of that topic in a cell.

To do this, `runModels()` builds several models (e.g. with diferent numbers of topics) using Latent Dirichlet Allocation (LDA) on the binary accessibility matrix (automatically stored in the initialized `cisTopicObject`). We can then select the best model using `selectModel()` and `logLikelihoodByIter()`.

The main parameters for running the models (`runModels`) are:

* **Number of topics** (`topic`): The number of topics are usually slightly bigger than the potential cell states in the data set. In the case of single cell epigenomics data the number of topics is low compared to other implementations (e.g. text classification). The running time will be affected by the number of topics.

* The Dirichlet hyperparameters **alpha** (`topic proportions`) and **beta** (`topic multinomials`): **Alpha** affects to the topic-cell contributions; a low alpha forces to pick for each cell a few topics with significant contribution, while a high alpha allows cells to have similar, smooth topic proportions. **Beta** affects to the region combinations; the lower the beta, the fewer regions a topic will have; the higher the beta, the less distinct these topics will be (i.e. there will be more overlap between the topics). By default, we select alpha as 50/number of topics and beta as 0.1 (as Griffiths & Steyvers, 2004).

* Number of **iterations** and **burnin**: For recording the assignments, it is necessary that the likelihood of the model is already stabilised.  cisTopic counts with the function `logLikelihoodByIter` to check whether this parameters should be changed. The number of iterations affect the speed of the algorithm. Note that the burnin will be substracted from the number of iterations.

**NOTE:** For large data sets it may not be feasible to keep all models simultaneously in memory. An alternative is to run the models and only save their likelihoods and the model with the highest likelihood (see the argument *returnType* in *runModels*). If after checking the likelihood plot another model is preferred, the function can be re-run only for that number of topics.

In this tutorial, we will test models with 2, 5, 10, 15, 20, 25, 30, 35 and 40 topics [Reference running time for the example data set (5k cells; ~97k regions): 80 min].

```{r, eval=F}
cisTopicObject <- runCGSModels(cisTopicObject, topic=c(2, 5, 10, 15, 20, 25, 30, 35, 40), seed=987, nCores=9, burnin = 20, iterations = 50, addModels=FALSE)
```

### Selection of the best model

The log likelihood can be used to estimate the plausibility of a model parameter value, given the observed data.
`selectModel` will select the model with the highest log likelihood (`P(D|T)`) at the last iteration.

```{r, fig.show='hold', fig.align='center'}
cisTopicObject <- selectModel(cisTopicObject)
```

If two or more models have comparable log likelihoods, we recommend to pick the one with the lower number of topics (i.e. lower complexity). By default, this function selects the model with the highest likelihood, but the user can select a certain topic with the `select` parameter in this function. In cases where the maximum does not seem to be achieved, the user can add additional models setting `addModels=TRUE`. [Reference running time for the example data set (5k cells; ~97k regions): 80 min].

Another way of visualizing the likelihood of the models is to plot their changes through the different iterations. It is important to check that the likelihood of the models is stabilised in the recording iterations, and the area under these curves can also be useful for model selection.

```{r, fig.show='hold', fig.align='center'}
logLikelihoodByIter(cisTopicObject, select=c(2,5,10,15,20,25,30,35,40))
```

If the models are stabilized after burnin (grey line), we can conclude that the selection of the `number of iterations` and `burnin` was suitable.

In this example, we will select 30 as the optimal number of topics.

```{r, fig.show='hold', fig.align='center'}
cisTopicObject <- selectModel(cisTopicObject, select=30)
```

## Interpreting the models

### A. Identification of cell states using the cell-cisTopic distributions

LDA returns two distributions that represent (1) the topic contributions per cell and (2) the region contribution to a topic. We can interpret these values as a dimensinality reduction method, after which the data is re-represented as a matrix with cells as columns, topics as rows and contributions as values. The recorded topic assignments to the cells (not normalised) are stored in `cisTopicObject@selected.model$document_expects` (see `lda` package).

Different methods can be used for clustering and/or visualization. cisTopic includes wrapper functions to easily run Umap, tSNE, diffussion maps and PCA (the results are saved in the slot `@dr$cell`):

```{r, eval=F}
cisTopicObject <- runtSNE(cisTopicObject, target='cell', seed=123, pca=F, method='Probability')
```

Additionally, we can use the cell-topic matrix to define clusters. The user can select the preferred methodology to do this. In this case, we will apply a peak density algorithm on the tsne dimensionality reduction projections, as in Cusanovich et al. (2018).

If you want to retrieve the normalised assignments for other analyses, you can use the function `modelMatSelection`:

```{r}
cellassign <- modelMatSelection(cisTopicObject, 'cell', 'Probability')
```

```{r}
set.seed(123)
library(Rtsne)
DR <- Rtsne(t(cellassign), pca=F)
DRdist <- dist(DR$Y)
library(densityClust)
dclust <- densityClust(DRdist,gaussian=T)
dclust <- findClusters(dclust, rho = 50, delta = 2.5)
```

```{r, fig.show='hold', fig.align='center'}
# Check thresholds
options(repr.plot.width=6, repr.plot.height=6)
plot(dclust$rho,dclust$delta,pch=20,cex=0.6,xlab='rho', ylab='delta')
points(dclust$rho[dclust$peaks],dclust$delta[dclust$peaks],col="red",pch=20,cex=0.8)
text(dclust$rho[dclust$peaks]-2,dclust$delta[dclust$peaks]+1.5,labels=dclust$clusters[dclust$peaks])
abline(v=50)
abline(h=2.5)
```

```{r, eval=F}
# Add cluster information
densityClust <- dclust$clusters
densityClust <- as.data.frame(densityClust)
rownames(densityClust) <- cisTopicObject@cell.names
colnames(densityClust) <- 'densityClust'
densityClust[,1] <- as.factor(densityClust[,1])
cisTopicObject <- addCellMetadata(cisTopicObject, densityClust)
```

Once calculations are done, cisTopic offers a unified visualization function (`plotFeatures`), which allows to visualize tSNE, diffussion maps, principal components and biplots (in 2/3D), colored by metadata and/or topic enrichment. 

```{r, fig.show='hold', fig.align='center'}
par(mfrow=c(1,2))
plotFeatures(cisTopicObject, method='tSNE', target='cell', topic_contr=NULL, colorBy=c('nCounts', 'nAcc','densityClust'), cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE, col.low='darkgreen', col.mid='yellow', col.high='brown1', intervals=10)
```

We can also generate a heatmap based on the cell-cisTopic distributions.

```{r, fig.show='hold', fig.align='center'}
cellTopicHeatmap(cisTopicObject, method='Probability', colorBy=c('densityClust'))
```

To color the tSNE by topic score:

```{r}
par(mfrow=c(2,5))
plotFeatures(cisTopicObject, method='tSNE', target='cell', topic_contr='Probability', colorBy=NULL, cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE)
```

#### Enrichment of epigenomic signatures in the cells

By multiplying the cell and topic assignments, the likelihood of each region in each cell (i.e. predictive distribution). This matrix is stored in `object@predictive.distribution`. These distributions can be used to estimate drop-outs and build cell-specific region rankings that can be used with `AUCell` for estimating the enrichment of epigenomic signatures within the cells.

```{r, eval=F}
pred.matrix <- predictiveDistribution(cisTopicObject)
```

For example, we can evaluate which cells are more enriched for certain ChIP-seq signatures. First, epigenomic regions are intersected and mapped to regions in the dataset (by default, with at least 40% overlap). To test the enrichment of these signatures in each cell, we use a GSEA-like recovery curve ranking-based approach.  In each cell, regions are ranked based on their probability (x-axis), and when a region is present in the signature we increase one step in the y-axis. The Area Under the Curve (AUC) is used to evaluate the importance of that signature within that cell. The corresponding overlapping sets (which are stored in `object@signatures`) are used as input, together with the cell-specific region rankings, for the function `signatureCellEnrichment`. AUC values for each specific signature are stored in `object@cell.data`. In this case, we can use bulk signatures from the hematopoietic system from Corces et al. (2016).

```{r, eval=F}
# Obtain signatures
path_to_signatures <- paste0(pathTo10X, 'Bulk_peaks/')
Bulk_ATAC_signatures <- paste(path_to_signatures, list.files(path_to_signatures), sep='')
labels  <- gsub('._peaks.narrowPeak', '', list.files(path_to_signatures))
cisTopicObject <- getSignaturesRegions(cisTopicObject, Bulk_ATAC_signatures, labels=labels, minOverlap = 0.4)
# To only keep unique peaks per signature
cisTopicObject@signatures <- llply(1:length(cisTopicObject@signatures), function (i) cisTopicObject@signatures[[i]][-which(cisTopicObject@signatures[[i]] %in% unlist(as.vector(cisTopicObject@signatures[-i])))]) 
names(cisTopicObject@signatures) <- labels
```

```{r, eval=F}
# Compute cell rankings (Reference time: 9 min)
library(AUCell)
aucellRankings <- AUCell_buildRankings(pred.matrix, plot=FALSE, verbose=FALSE)
```

```{r, eval=F}
# Check signature enrichment in cells (Reference time: 1 min)
cisTopicObject <- signatureCellEnrichment(cisTopicObject, aucellRankings, selected.signatures='all', aucMaxRank = 0.3*nrow(aucellRankings), plot=FALSE)
```

```{r, fig.show='hold', fig.align='center'}
# Plot
par(mfrow=c(2,2))
plotFeatures(cisTopicObject, method='tSNE', target='cell', topic_contr=NULL, colorBy=c('CD4Tcell', 'Mono', 'Bcell', 'NKcell'), cex.legend = 0.4, factor.max=.75, dim=2, legend=TRUE, intervals=10)
```

**NOTE**: The predictive distributions and the AUCell rankings are not stored in the cisTopic object as they have a big size.

### B. Analysis of the regulatory topics

#### Defining topics

To analyze the regions included in the cisTopics, the first step is always to derive a score that evaluates how likely is for a region to belong to a topic. `getRegionsScores()` calculates these scores based on the proportion of region specific assignments to a topic. These scores can be rescaled into the range [0,1], which will be useful for the binarization step (as it will force data to follow a gamma distribution shape). This information is stored in the `region.data` slot.

```{r, eval = T}
cisTopicObject <- getRegionsScores(cisTopicObject, method='NormTop', scale=TRUE)
```


#### Topic visualization

Based on the topic scores for each region, different methods can be used for clustering and/or visualization (as shown for cells). cisTopic includes wrapper functions to easily run Umap, tSNE, diffussion maps and PCA (the results are saved in the slot `@dr$region`). In the case of regions, only high confidence regions (i.e. that pass the binarization threshold at least in 1 topic) are used:

```{r, eval=F}
par(mfrow=c(2,5))
cisTopicObject <- binarizecisTopics(cisTopicObject, thrP=0.975, plot=TRUE)
cisTopicObject <- runtSNE(cisTopicObject, target='region', perplexity=200, check_duplicates=FALSE)
```

The function `plotFeatures` can also be used to visualize region-based tSNEs, diffussion maps, principal components and biplots (in 2/3D), colored by metadata and/or topic enrichment. 

```{r, fig.show='hold', fig.align='center'}
plotFeatures(cisTopicObject, method='tSNE', target='region', topic_contr=NULL, colorBy=c('nCells'), cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE, col.low='darkgreen', col.mid='yellow', col.high='brown1', intervals=10)
par(mfrow=c(2,5))
plotFeatures(cisTopicObject, method='tSNE', target='region', topic_contr='Z-score', colorBy=NULL, cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE, col.low='darkgreen', col.mid='yellow', col.high='brown1')
```



```{r, eval=F}
library(org.Hs.eg.db)
cisTopicObject <- annotateRegions(cisTopicObject, txdb=TxDb.Hsapiens.UCSC.hg19.knownGene, annoDb='org.Hs.eg.db')
```


#####  Gene accessibility scores

We can also evaluate the overall accessibility around certain markers using the predictive distribution. In this case, we will sum the probability of each region linked to each marker gene (based on ChIPseeker annotations in this tutorial).

```{r, eval=F}
region2gene <- cisTopicObject@region.data[,'SYMBOL', drop=FALSE]
region2gene <- split(region2gene, region2gene[,'SYMBOL']) 
region2gene <- lapply(region2gene, rownames) 
# From pretrained Garnett's model markers on pBMC dataset from 10X (Pliner et al, 2019)
selectedGenes <- c('CD34', 'THY1', 'ENG', 'KIT', 'PROM1', #CD34+
                   'NCAM1', 'FCGR3A', #NKcells
                   'CD14', 'FCGR1A', 'CD68', 'S100A12', #Monocytes
                   'CD19', 'MS4A1', 'CD79A', #Bcells
                   'CD3D', 'CD3E', 'CD3G', #Tcells
                   'CD4', 'FOXP3', 'IL2RA', 'IL7R', #CD4 Tcell
                   'CD8A', 'CD8B', #CD8 Tcell
                   'IL3RA', 'CD1C', 'BATF3', 'THBD', 'CD209' #Dendritic cells
                   )
region2gene_subset <- region2gene[which(names(region2gene) %in% selectedGenes)]
predMatSumByGene <- sapply(region2gene_subset, function(x) apply(pred.matrix[x,, drop=F], 2, sum))
rownames(predMatSumByGene) <- cisTopicObject@cell.names
# Add to cell data
cisTopicObject <- addCellMetadata(cisTopicObject, predMatSumByGene)
```

```{r, fig.show='hold', fig.align='center'}
par(mfrow=c(1,2))
plotFeatures(cisTopicObject, method='tSNE', target='cell', topic_contr=NULL, colorBy=c('KIT', 'PROM1', 'NCAM1', 'FCGR3A','CD14','S100A12', 'MS4A1', 'CD79A', 'CD3D', 'CD3E', 'CD4', 'CD8A'), cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE, intervals=10)
```

Finally, you can save your cisTopic object:

```{r, eval = FALSE}
saveRDS(cisTopicObject, file='cisTopicObject_pbmc.Rds')
```

# References

1. Blei, D. M., Ng, A. Y., & Jordan, M. I. (2003). Latent dirichlet allocation. *Journal of machine Learning research*, 3(Jan), 993-1022.
2. Steyvers, M., & Griffiths, T. (2007). Probabilistic topic models. *Handbook of latent semantic analysis*, 427(7), 424-440.
3. Cusanovich, D. A., Reddington, J. P., Garfield, D. A., Daza, R. M., Aghamirzaie, D., Marco-Ferreres, R., ... & Trapnell, C. (2018). The cis-regulatory dynamics of embryonic development at single-cell resolution. *Nature*, 555(7697), 538.
4. Aibar, S., Bravo GonzÃ¡lez-Blas, C., Moerman, T., Imrichova, H., Hulselmans, G., Rambow, F., ... & Atak, Z. K. (2017). SCENIC: single-cell regulatory network inference and clustering. *Nature methods*, 14(11), 1083.
5. Medina-Rivera, A., Defrance, M., Sand, O., Herrmann, C., Castro-Mondragon, J. A., Delerce, J., ... & Staines, D. M. (2015). RSAT 2015: regulatory sequence analysis tools. *Nucleic acids research*, 43(W1), W50-W56.
6. Heinz, S., Benner, C., Spann, N., Bertolino, E., Lin, Y. C., Laslo, P., ... & Glass, C. K. (2010). Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities. *Molecular cell*, 38(4), 576-589.
7. Chiu, T. P., Yang, L., Zhou, T., Main, B. J., Parker, S. C., Nuzhdin, S. V., ... & Rohs, R. (2014). GBshape: a genome browser database for DNA shape annotations. *Nucleic acids research*, 43(D1), D103-D109.
8. Pliner, H. A., Shendure, J., & Trapnell, C. (2019). Supervised classification enables rapid annotation of cell atlases. *BioRxiv*, 538652.


# Session Infomation

```{r, echo=T,eval=F,message=F,warning=F,error=T}
sessionInfo()
```

---
*The lesson material was developed by members of the BMBL.*
