---
title: "Heatmap"
author: "Cankun Wang"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
library(here)
library(qs)
library(tidyverse)
library(Seurat)
library(SeuratDisk)
```


# Set working dir

```{r,echo=F,eval=T,message=F,warning=F,error=F}
set.seed(1)

here::i_am("heatmap.rmd")
print(paste("Current working directory:", here::here()))
source("../common/functions.R")

```

```{r,echo=F,eval=T,message=F,warning=F,error=F}
combined <- qs::qread("example_combined.qsave")

```

# Heatmap using Seurat

```{r, fig.width=10, fig.height=4}

DefaultAssay(combined) <- "RNA"
two_color <- c('#C0C0C0', '#B00D23')

this_genes <-
  c(
    "CD14",
    "LYZ",
    "FCGR3A",
    "MS4A7",
    "MARCO",
    "ITGAM",
    "ADGRE1",
    "FCER1A",
    "CST3",
    "IL3RA",
    "GZMB",
    "CD79A",
    "MS4A1",
    "CD3D",
    "IL7R",
    "CCR7",
    "CD8A",
    "GNLY",
    "NKG7"
  )

mapal <- colorRampPalette(RColorBrewer::brewer.pal(6, "RdBu"))(128)

combined <- ScaleData(combined, features = this_genes)
p1 <-
  DoHeatmap(combined,
            features = this_genes,
            angle = 90,
            label = F) + scale_fill_gradientn(colours = rev(mapal))

print(p1)


```

## Save to PNG 

```{r,echo=T,eval=T, fig.height=4, fig.width=8}

png(
  paste0('./img/heatmap_cell_type_seurat.png'),
  width = 2500,
  height = 1200,
  res = 300
)
print(p1)
dev.off()

```



# Heatmap using pheatmap

```{r, fig.width=8, fig.height=10}

heatmap_matrix <-
  as.data.frame(t(as.matrix(combined@assays$RNA@data))) %>%
  mutate(cluster = as.character(combined$cell_type)) %>%
  group_by(cluster) %>%
  summarise_at(vars(this_genes), mean) %>%
  ungroup() %>%
  as.data.frame() %>%
  column_to_rownames('cluster')

pheatmap::pheatmap(
  t(heatmap_matrix),
  scale = "row",
  color = colorRampPalette(c(
    "#344298", "#bbe0ed", "#fdf9b7", "#fba95e", "#ae0825"
  ))(1000),
  cluster_rows = F,
  cluster_cols = T,
  angle_col = 45,
  cellheight = 16,
  cellwidth = 16
)
```

## Save to PNG 

```{r,echo=T,eval=T, fig.height=4, fig.width=8}

png(
  paste0('./img/heatmap_marker_pheatmap.png'),
  width = 1600,
  height = 2000,
  res = 300
)

pheatmap::pheatmap(
  t(heatmap_matrix),
  scale = "row",
  color = colorRampPalette(c(
    "#344298", "#bbe0ed", "#fdf9b7", "#fba95e", "#ae0825"
  ))(1000),
  cluster_rows = F,
  cluster_cols = T,
  angle_col = 45,
  cellheight = 16,
  cellwidth = 16
)

dev.off()

```

# Dot plot using plot1cell package

More examples: https://github.com/TheHumphreysLab/plot1cell

## How to install plot1cell package

```{r, eval=T}

bioc_packages <- c("biomaRt", "GenomeInfoDb","EnsDb.Hsapiens.v86","GEOquery","simplifyEnrichment","ComplexHeatmap")

check_plot1cell <-
  !("plot1cell" %in% installed.packages()[, "Package"])
if (check_plot1cell) {
  bioc_packages <-
    c(
      "biomaRt",
      "GenomeInfoDb",
      "EnsDb.Hsapiens.v86",
      "GEOquery",
      "simplifyEnrichment",
      "ComplexHeatmap"
    )
  dev_packages <- c("DoubletFinder", "hdf5r", "loomR")
  dev.packages <-
    c("chris-mcginnis-ucsf/DoubletFinder",
      "Novartis/hdf5r",
      "mojaveazure/loomR")
  bioc_np <-
    bioc_packages[!(bioc_packages %in% installed.packages()[, "Package"])]
  dev_np <-
    dev_packages[!(dev_packages %in% installed.packages()[, "Package"])]
  
  if (!require("BiocManager"))
    install.packages("BiocManager")
  if (length(bioc_np)) {
    BiocManager::install(bioc_np)
  }
  if (length(dev_np)) {
    devtools::install_github(dev.packages[!(dev_packages %in% installed.packages()[, "Package"])])
  }
  devtools::install_github("TheHumphreysLab/plot1cell")
}

```


## ComplexHeatmap to show unique genes across groups/cell types

**plot1cell can directly identify the condition specific genes in a selected cell type/group and plot those genes using ComplexHeatmap**

```{r, fig.width=5, fig.height=10}
library(plot1cell)
DefaultAssay(combined) <- "RNA"

this_genes <-
  c(
    "FCGR3A",
    "LYZ",
    "MS4A7",
    "MARCO",
    "ITGAM",
    "ADGRE1",
    "FCER1A",
    "CST3",
    "IL3RA",
    "GZMB",
    "CD79A",
    "MS4A1",
    "CD3D",
    "IL7R",
    "CCR7",
    "CD8A",
    "GNLY",
    "NKG7"
  )

tmp_ident <- as.factor(combined$orig.ident)
levels(tmp_ident) <- rep("Gene expression", length(levels(tmp_ident)))
combined <- AddMetaData(combined, tmp_ident, col.name = "gene_expression")
Idents(combined) <- combined$gene_expression

combined <- ScaleData(combined, features = rownames(combined))
complex_heatmap_unique(
  seu_obj = combined,
  celltype = "Gene expression",
  group = "cell_type",
  gene_highlight = this_genes
)

```

![heatmap_plot1cell_cell_type_markers.png](./img/heatmap_plot1cell_cell_type_markers.png)

## Save to PNG 

```{r}

png(
  paste0('./img/heatmap_plot1cell_cell_type_markers.png'),
  width = 1800,
  height = 4000,
  res = 300
)
complex_heatmap_unique(
  seu_obj = combined,
  celltype = "Gene expression",
  group = "cell_type",
  gene_highlight = this_genes
)

dev.off()

```

