---
title: "ST_general_workflow_tutorial"
author: "YOUR NAME"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


This R Markdown tutorial was created using resources provided by the Satija Lab (https://satijalab.org/seurat/articles/spatial_vignette.html). It utilizes Seurat (>=3.2) to analyze spatially-resolved RNA-seq data. 

# Load Seurat and other packages
```{r}
library(Seurat)
library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
```

This tutorial will use a 10x Genomics dataset that is accessible through the `SeuratData` package.

# Load data
```{r}
InstallData("stxBrain")
brain <- LoadData("stxBrain", type = "anterior1")
```

# Preprocessing
The first step of data preprocessing is Normalization.
```{r}
plot1 <- VlnPlot(brain, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
plot2 <- SpatialFeaturePlot(brain, features = "nCount_Spatial") + theme(legend.position = "right")
wrap_plots(plot1, plot2)

brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
```

#Gene expression visualization
The gene Hpca in the data set used in the tutorial is a strong hippocampus marker and Ttr is a marker of the choroid plexus. Molecular data can be overlaid on top of tissue histology to show this.

```{r}
SpatialFeaturePlot(brain, features = c("Hpca", "Ttr"))
```

"pt.size.factor" can be used to scale the size of the spots. "alpha" can be used to minimize or maximize transparency. 

```{r}
p1 <- SpatialFeaturePlot(brain, features = "Ttr", pt.size.factor = 1)
p2 <- SpatialFeaturePlot(brain, features = "Ttr", alpha = c(0.1, 1))
p1 + p2
```

# Dimensionality reduction, clustering, and visualization
The same workflow that is used for scRNA-seq analysis can be applied here.

```{r}
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:30)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:30)
```

Results can be visualized using UMAP or overlaid on the image.

```{r}
p1 <- DimPlot(brain, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(brain, label = TRUE, label.size = 3)
p1 + p2
```

The `cells.highlight`  paramter and `SpatialDimPlot()` function can be used to note cells of interest to distinguish clusters' spatial localization.

```{r}
SpatialDimPlot(brain, cells.highlight = CellsByIdentities(object = brain, idents = c(2, 1, 4, 3,
    5, 8)), facet.highlight = TRUE, ncol = 3)
```

# Interactive plotting

When the `interactive` parameter in `SpatialDimPlot()` and `SpatialFeaturePlot()` is set to `TRUE`, the Rstudio viewer pane will open with an interactive Shiny plot.

```{r}
SpatialDimPlot(brain, interactive = TRUE)

SpatialFeaturePlot(brain, features = "Ttr", interactive = TRUE)

LinkedDimPlot(brain)
```

# Identification of Spatially Variable Features

Because the clusters identified using this example datat exhibit clear spatial resolution, we can perform differential expression based on pre-annotated anatomical regions in tissue.
```{r}
de_markers <- FindMarkers(brain, ident.1 = 5, ident.2 = 6)
SpatialFeaturePlot(object = brain, features = rownames(de_markers)[1:3], alpha = c(0.1, 1), ncol = 3)
```

An alternative approach is to search for features exhibiting spatial patterning without pre-annotation. 

```{r}
brain <- FindSpatiallyVariableFeatures(brain, assay = "SCT", features = VariableFeatures(brain)[1:1000],
    selection.method = "markvariogram")

top.features <- head(SpatiallyVariableFeatures(brain, selection.method = "markvariogram"), 6)
SpatialFeaturePlot(brain, features = top.features, ncol = 3, alpha = c(0.1, 1))
```

#Subset out anatomical regions

For this data, we will subset the frontal cortex. The first step is to take a subset of cluster and further segment based on exact positions.
```{r}
cortex <- subset(brain, idents = c(1, 2, 3, 4, 6, 7))
# now remove additional cells, use SpatialDimPlots to visualize what to remove
# SpatialDimPlot(cortex,cells.highlight = WhichCells(cortex, expression = image_imagerow > 400
# | image_imagecol < 150))
cortex <- subset(cortex, anterior1_imagerow > 400 | anterior1_imagecol < 150, invert = TRUE)
cortex <- subset(cortex, anterior1_imagerow > 275 & anterior1_imagecol > 370, invert = TRUE)
cortex <- subset(cortex, anterior1_imagerow > 250 & anterior1_imagecol > 440, invert = TRUE)
```

We can visualize the cortical cells next.

```{r}
p1 <- SpatialDimPlot(cortex, crop = TRUE, label = TRUE)
p2 <- SpatialDimPlot(cortex, crop = FALSE, label = TRUE, pt.size.factor = 1, label.size = 3)
p1 + p2
```

# Integration with single-cell data
The single-cell data can be downloaded here (https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1). We must pre-process the scRNA-seq reference and perform label transfer.

```{r}
allen_reference <- readRDS("../data/allen_cortex.rds")
```

```{r}
# note that setting ncells=3000 normalizes the full dataset but learns noise models on 3k
# cells this speeds up SCTransform dramatically with no loss in performance
library(dplyr)
allen_reference <- SCTransform(allen_reference, ncells = 3000, verbose = FALSE) %>%
    RunPCA(verbose = FALSE) %>%
    RunUMAP(dims = 1:30)
```

```{r}
# After subsetting, we renormalize cortex
cortex <- SCTransform(cortex, assay = "Spatial", verbose = FALSE) %>%
    RunPCA(verbose = FALSE)
# the annotation is stored in the 'subclass' column of object metadata
DimPlot(allen_reference, group.by = "subclass", label = TRUE)
```

```{r}
anchors <- FindTransferAnchors(reference = allen_reference, query = cortex, normalization.method = "SCT")
predictions.assay <- TransferData(anchorset = anchors, refdata = allen_reference$subclass, prediction.assay = TRUE,
    weight.reduction = cortex[["pca"]], dims = 1:30)
cortex[["predictions"]] <- predictions.assay
```

Next, distinguish between distinct sequential layers of neuronal subtypes.

```{r}
DefaultAssay(cortex) <- "predictions"
SpatialFeaturePlot(cortex, features = c("L2/3 IT", "L4"), pt.size.factor = 1.6, ncol = 2, crop = TRUE)
```

Now, we will predict cell types with spatially restricted locations.

```{r}
cortex <- FindSpatiallyVariableFeatures(cortex, assay = "predictions", selection.method = "markvariogram",
    features = rownames(cortex), r.metric = 5, slot = "data")
top.clusters <- head(SpatiallyVariableFeatures(cortex), 4)
SpatialPlot(object = cortex, features = top.clusters, ncol = 2)
```

Show that the procedure can recover known spatial localization patterns of neuronal and non-neuronal subsets.

```{r}
SpatialFeaturePlot(cortex, features = c("Astro", "L2/3 IT", "L4", "L5 PT", "L5 IT", "L6 CT", "L6 IT",
    "L6b", "Oligo"), pt.size.factor = 1, ncol = 2, crop = FALSE, alpha = c(0.1, 1))
```

# Working with multiple slices in Seurat

Read and perform initial normalization of additional mouse brain slice from the dataset.

```{r}
brain2 <- LoadData("stxBrain", type = "posterior1")
brain2 <- SCTransform(brain2, assay = "Spatial", verbose = FALSE)
```

Work with multiple slices in the same Seurat object:
```{r}
brain.merge <- merge(brain, brain2)
```

Perform joint dimensional reduction and clustering on the underlying RNA expression data.

```{r}
DefaultAssay(brain.merge) <- "SCT"
VariableFeatures(brain.merge) <- c(VariableFeatures(brain), VariableFeatures(brain2))
brain.merge <- RunPCA(brain.merge, verbose = FALSE)
brain.merge <- FindNeighbors(brain.merge, dims = 1:30)
brain.merge <- FindClusters(brain.merge, verbose = FALSE)
brain.merge <- RunUMAP(brain.merge, dims = 1:30)
```

Visualize the data in a UMAP plot.

```{r}
DimPlot(brain.merge, reduction = "umap", group.by = c("ident", "orig.ident"))

SpatialDimPlot(brain.merge)

SpatialFeaturePlot(brain.merge, features = c("Hpca", "Plp1"))
```


```{r}
sessionInfo()
```
---
*This lesson materials was originally created by the Satija Lab, and has been developed by members of the BMBL.*

